### PR TITLE
Move database fetchers to discovery service

### PR DESCRIPTION
**Pull request by greedy52**
Implements #18742
- #18742

Sorry for going back and forth on this. Decided to do the proper move after some discussion from the previous PR.

Note that this change is only the moving/refactoring part. Will enable database fetchers in discovery service in a follow-up PR (redo of #19030 on top of this change). So far, no existing functionalities are expected to change.

Overview of changes:
- Moved `lib/srv/discovery/fetcher/fetcher.go` to `lib/srv/discovery/common/interfaces.go`
- Combined db cloud watcher and discovery kube watcher logic to a new common watcher in `lib/srv/discovery/common/watcher.go` (which is the only "logic" change)
- Moved database fetchers from `lib/srv/db/cloud/watchers/` to `lib/srv/discovery/fetchers/db/`
- Moved cloud client mocks from `lib/srv/db/cloud/mocks.go` to `lib/cloud/test/`
- Split big old `watcher_test.go` to individual tests (which is the a LOT changes)
- Updated database service and discovery kube watcher to use the new common watcher.

### CODE REVIEW

File: lib/services/matchers.go [hunk 1]
```diff
@@ -88,6 +91,40 @@ type GCPMatcher struct {
...
+			subs = []string{types.Wildcard}
+		}
+		if len(groups) == 0 || slices.Contains(groups, types.Wildcard) {
+			groups = []string{types.Wildcard}
+		}
+		if len(regions) == 0 || slices.Contains(regions, types.Wildcard) {
+			regions = []string{types.Wildcard}
+		} else {
+			for i, region := range regions {
+				regions[i] = azureutils.NormailizeLocation(region)
```
Comment (nklaassen) [inline lib/services/matchers.go:114]:
```suggestion
				regions[i] = azureutils.NormalizeLocation(region)
```

File: lib/srv/discovery/fetchers/db/aws_rds_proxy_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			name:          "fetch vpc1",
+			inputLabels:   map[string]string{"vpc-id": "vpc1"},
+			wantDatabases: types.Databases{rdsProxyDatabaseVpc1, rdsProxyEndpointDatabaseVpc1},
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
```
Comment (nklaassen) [inline lib/srv/discovery/fetchers/db/aws_rds_proxy_test.go:67]: I noticed you've parallelized the subtests but not the top level tests in this package, is that by design? I think this way none of the setup code runs in parallel

**Review by nklaassen — APPROVED** [review]

File: lib/srv/discovery/fetchers/db/aws_rds_proxy_test.go (hunk 2 above)
Comment (greedy52) [reply lib/srv/discovery/fetchers/db/aws_rds_proxy_test.go:69]: To be honest I didn't give it too much thought. The tests in this lib took about half a second with or without parallel. The setup codes are very quick (just making mock structs) but the test cases are pretty quick too. Anyway, added parrallel to all tests for best parrallelization.

File: lib/cloud/test/aws.go [hunk 3]
```diff
@@ -14,11 +14,9 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package cloud
```
Comment (tigrato) [inline lib/cloud/test/aws.go:17]: I think this package should be renamed to `mocks` 
It scopes the logic better than `test`

File: lib/srv/db/access_test.go [hunk 4]
```diff
@@ -1950,6 +1955,20 @@ func (p *agentParams) setDefaults(c *testContext) {
 			},
 		}
 	}
+
+	if p.CloudClients == nil {
+		p.CloudClients = &clients.TestCloudClients{
+			STS:                &cloudtest.STSMock{},
```
Comment (tigrato) [inline lib/srv/db/access_test.go:1961]: this can be wrapped in a `cloudtest.New` that fills the default values

**Review by tigrato — APPROVED** [review]

File: lib/cloud/test/aws.go (hunk 3 above)
Comment (greedy52) [reply lib/cloud/test/aws.go]: Ya, we had a small discussion in the previous PR and it does appear it's more a `mocks`. Updated now.

File: lib/srv/db/access_test.go (hunk 4 above)
Comment (greedy52) [reply lib/srv/db/access_test.go]: It's not very common to mock all the clients like this. So will keep this one as it is (being lazy) for now.

**Pull request merged**